### PR TITLE
Include conv layers with same padding from MUGEN

### PR DESCRIPTION
**Pull request by RdoubleA**
Summary:
Ported `SamePadConv3D` layers from MUGEN codebase with some redesign to apply to more input dim/kernel/stride cases. Small changes for mypy and an added rule to ignore for flake8.

Test plan:
Added unit tests. Local testing on notebook to compare outputs from original MUGEN implementation and ours to make sure they are exact same shape and value.

`SamePadConvTranspose3d`
image

`SamePadConv3d`
image

`pytest -vv`
image

### CODE REVIEW

File: torchmultimodal/modules/layers/conv.py [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+        out_channels: int,
+        kernel_size: _size_3_t,
+        stride: _size_3_t = 1,
+        bias: bool = True,
+        **kwargs
+    ) -> None:
+        super().__init__()
+
+        self.pad_input: Tuple = None
+        self.kernel_size = _triple(kernel_size)
```
[the remark this answers is not in the document]
Comment (RdoubleA) [reply torchmultimodal/modules/layers/conv.py:45]: mainly used it because I had to do this int/tuple handling in multiple locations, but I can make it more explicit in favor of readability

File: torchmultimodal/modules/layers/conv.py [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+        self.stride = _triple(stride)
+
+        self.conv = nn.ConvTranspose3d(
+            in_channels, out_channels, kernel_size, stride=stride, bias=bias, **kwargs
+        )
+
+    def forward(self, x) -> Tensor:
```
[the remark this answers is not in the document]
Comment (RdoubleA) [reply torchmultimodal/modules/layers/conv.py:103]: why does mypy miss these?

File: torchmultimodal/modules/layers/conv.py [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+        # Calculate padding needed based on input shape only once to reduce run time
+        if self.pad_input is None:
+            self.pad_input = calculate_same_padding(
+                self.kernel_size, self.stride, x.shape[2:]
+            )
+            self.conv.padding, self.conv.output_padding = calculate_transpose_padding(
```
[the remark this answers is not in the document]
Comment (RdoubleA) [reply torchmultimodal/modules/layers/conv.py:109]: pad_input won't be affected by the padding argument, it should still be none and it will reset the padding as intended agnostic of what the user feeds in. but yeah I agree, let me see if I can make this more explicit

File: torchmultimodal/modules/layers/conv.py [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+    """
+
+    convert_to_tuple = _ntuple(len(input_shape))
+    kernel_size = convert_to_tuple(kernel_size)
+    stride = convert_to_tuple(stride)
+
+    if not (len(kernel_size) == len(stride) == len(input_shape)):
+        raise ValueError("dims for kernel, stride, and input must match")
+    input_pad = _ntuple(len(input_shape) * 2)(input_pad)
+    if len(input_pad) % 2 != 0 or len(input_pad) // 2 != len(input_shape):
```
[the remark this answers is not in the document]
Comment (RdoubleA) [reply torchmultimodal/modules/layers/conv.py:193]: it would fail if input_pad is already a tuple, which ntuple will not touch. then if the input_pad tuple is not the right length the assertion with fail. I will probably remove the use of PyTorch specific data types and remove ntuple, that way it's much clearer

File: test/modules/layers/test_conv.py [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+class TestSamePadConv3d(unittest.TestCase):
+    """
+    Test the SamePadConv3d class and associated helpers
+    """
+
+    def setUp(self):
+        inputs = [torch.ones(1, 1, 8, 8, 8), torch.ones(1, 1, 7, 7, 7)]
+        kernels = [(4, 4, 4), (3, 3, 3)]
+        strides = [(2, 2, 2), (3, 3, 3)]
+        self.test_cases = list(product(*[inputs, kernels, strides]))
```
[the remark this answers is not in the document]
Comment (RdoubleA) [reply test/modules/layers/test_conv.py:29]: so I do want to test all these test cases, as they are each important (even vs odd input size, even vs odd kernel, even vs odd stride). I could list out all the combinations manually since there should be only 8 cases, I was initially thinking it wouldn't be much more readable seeing multiple long lists. let me make the change and we can reevaluate

[the remark this answers is not in the document]
Comment (ebsmothers) [reply test/modules/layers/test_conv.py:29]: Sounds good. I actually don't have an especially strong stance here, if listing them all out leads to a ton of boilerplate code it's also fine to leave as it is.

File: torchmultimodal/modules/layers/conv.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ebsmothers) [reply torchmultimodal/modules/layers/conv.py:103]: So I think we need to set `disallow_untyped_defs = True` in our mypy.ini file. I just ran it and it found 41 errors though. Maybe a nice BE task to clean all these up...

[the remark this answers is not in the document]
Comment (RdoubleA) [reply torchmultimodal/modules/layers/conv.py:103]: made a task: [T122368787](https://www.internalfb.com/intern/tasks/?t=122368787)

File: test/modules/layers/test_conv.py [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+            _ = calculate_transpose_padding((3, 3), (2, 2), (5, 5), (1, 0, 1))
+            _ = calculate_transpose_padding(3, 2, (5, 5, 5), (1, 1, 1, 1, 1, 1, 1))
+
+    def test_calculate_transpose_padding_output(self):
+        for i, (inp, kernel, stride) in enumerate(self.test_cases):
+            pad = calculate_same_padding(kernel, stride, inp.shape[2:])
+            transpose_pad_actual, output_pad_actual = calculate_transpose_padding(
+                kernel, stride, inp.shape[2:], pad
+            )
+            self.assertEqual(
```
Comment (langong347) [inline test/modules/layers/test_conv.py:114]: any reason here we need to use self assertion?

File: test/modules/layers/test_conv.py [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+
+    def test_calculate_same_padding_assert(self):
+        with self.assertRaises(ValueError):
+            _ = calculate_same_padding((3, 3), (2, 2, 2), (5, 5))
+            _ = calculate_same_padding(3, (2, 2), (5, 5, 5))
+
+    def test_calculate_same_padding_output(self):
+        for i, (inp, kernel, stride) in enumerate(self.test_cases):
+            pad_actual = calculate_same_padding(kernel, stride, inp.shape[2:])
+            assert (
```
Comment (langong347) [inline test/modules/layers/test_conv.py:89]: Any reason here we use `assert` not `assert_expected`?

File: .flake8 [hunk 8]
```diff
@@ -1,11 +1,12 @@
...
-# Suggested config from pytorch that we can adapat
+# Suggested config from pytorch that we can adapt
 select = B,C,E,F,N,P,T4,W,B9
 max-line-length = 120
 # C408 ignored because we like the dict keyword argument syntax
 # E501 is not flexible enough, we're using B950 instead
+# N812 ignored because import torch.nn.functional as F is PyTorch convention
 ignore =
-    E203,E305,E402,E501,E721,E741,F405,F821,F841,F999,W503,W504,C408,E302,W291,E303,
+    E203,E305,E402,E501,E721,E741,F405,F821,F841,F999,W503,W504,C408,E302,W291,E303,N812,
```
Comment (langong347) [inline .flake8:9]: Going forward, you can break unrelated changes into a separate commit and use `ghstack` or `git worktree` to open stacked PRs.

File: torchmultimodal/modules/layers/conv.py [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+            )
+
+        self.conv = nn.ConvTranspose3d(
+            in_channels, out_channels, kernel_size, stride=stride, bias=bias, **kwargs
+        )
+
+    def forward(self, x: Tensor) -> Tensor:
+        # Calculate padding needed based on input shape only once to reduce run time
+        if self.pad_input is None:
+            self.pad_input = calculate_same_padding(
```
Comment (langong347) [inline torchmultimodal/modules/layers/conv.py:118]: Q: Since this is used in the VQVAE decoder where `x` should have been convolved already by the encoder,  why we need to pad `x` before "inflating" it back (i.e., transpose-conv)?

**Comment by langong347** [comment]
How about matching MUGEN when kernel = 3 ,stride = 2?

https://github.com/mugen-org/MUGEN_baseline/blob/main/lib/models/video_vqvae/vqvae.py#L344

File: torchmultimodal/modules/layers/conv.py [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+    to have the same shape as input, with stride 2 the dimensions of output are doubled.
+
+    The 'padding' argument in ConvTranspose effectively trims the output, and the 'output_padding'
+    argument effectively expands the output. These two knobs are adjusted to meet desired output dim.
+
+    Args:
+        kernel_size (int or Tuple): size of convolutional kernel
+        stride (int or Tuple): stride amount of kernel
+        input_shape (Tuple or Size): tuple describing shape of input, without batch or channel dimension
+        input_pad (int or Tuple): amount of padding added to input, must be twice length of kernel/stride/input_shape
```
[the remark this answers is not in the document]
Comment (RdoubleA) [reply torchmultimodal/modules/layers/conv.py:194]: resolved offline

File: .flake8 (hunk 8 above)
Comment (RdoubleA) [reply .flake8:9]: still need to learn ghstack...

File: test/modules/layers/test_conv.py (hunk 6 above)
Comment (RdoubleA) [reply test/modules/layers/test_conv.py:114]: `calculate_transpose_padding` returns Tuple, so I couldn't use `assert_expected`. I could cast back to `torch.tensor`, but I needed a more detailed message anyway so I stuck with `self.assertEqual`.

File: torchmultimodal/modules/layers/conv.py (hunk 9 above)
Comment (RdoubleA) [reply torchmultimodal/modules/layers/conv.py:118]: great question, which is what I was also wondering. currently consulting our partners in MUGEN but I'd say this is not a blocker for now

**Review by langong347 — APPROVED** [review]
